cpp/include/cudf/utilities/span.hpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/detail/raw_pointer_cast.h>
+#include <thrust/device_vector.h>
+
+#include <cstddef>
+#include <limits>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+
+constexpr std::size_t dynamic_extent = std::numeric_limits<std::size_t>::max();
+
+/**
+ * @brief C++20 std::span with reduced feature set.
+ */
+template <typename T, std::size_t Extent, typename Derived>
+class span_base {
+  static_assert(Extent == dynamic_extent, "Only dynamic extent is supported");
+
+ public:
+  using element_type    = T;
+  using value_type      = std::remove_cv<T>;
+  using size_type       = std::size_t;
+  using difference_type = std::ptrdiff_t;
+  using pointer         = T*;
+  using iterator        = T*;
+  using const_pointer   = T const*;
+  using reference       = T&;
+  using const_reference = T const&;
+
+  static constexpr std::size_t extent = Extent;
+
+  constexpr span_base() noexcept : _data(nullptr), _size(0) {}
+  constexpr span_base(pointer data, size_type size) : _data(data), _size(size) {}
+  // constexpr span_base(pointer begin, pointer end) : _data(begin), _size(end - begin) {}
+  constexpr span_base(span_base const& other) noexcept = default;
+  constexpr span_base& operator=(span_base const& other) noexcept = default;
+
+  template <typename C,
+            typename P = decltype(std::declval<C>().data()),
+            std::enable_if<std::is_same<std::add_pointer<T>, P>::value>* = nullptr>
```

Comment:
There's a few things problematic with this. First of all, it lets you do this:
```
std::vector<int> host_vector(N);
device_span d_s(host_vector); // This should not be allowed! 
```

Second of all, I'm surprised this works at all with `thrust::device_vector/host_vector`.

When `C` is `device_vector<T>`, then `P` will be `device_ptr<T>` and `is_same<T*, device_ptr<T>>` should fail.

Reply:
I've spent a fair amount of time attempting to make this work as a constructor, rather than type inferred `make_span(...)` factories. I imagine I'd eventually get the templated constructor to work, but even with examples like https://github.com/tcbrindle/span/blob/master/include/tcb/span.hpp and https://github.com/NVIDIA/libcudacxx/blob/main/libcxx/include/span, I've been unable to do so thus far. I will probably need some hands-on help. If it can wait until a followup PR, that would be preferable. If the generic constructor just needs to be deleted, and we can come back to supporting containers later, I can do that as well.

Reply:
What are you proposing for supporting ctors for containers for this PR then?

Reply:
This is one time I don't have a strong opinion. Template constructors are a stretch goal for me. Primarily I'm interested in cleaning up separate  `(..., ptr, size, ...)` args sourced from `rmm::device_(u)vector`, and we can clean that up without container-specific constructors. How can we get this PR cleaned up enough to merge?

Reply:
I don't think this PR should be merged unless you can do the following:
```
template <typename T>
__global__ kernel(device_span<T> s);
...

thrust::device_vector<T> v;
kernel<<<...>>>(v);
```
i.e., you can implicitly convert from the known device containers to the `device_span` type. The usefulness of `device_span` is significantly diminished if you have to do this:
```
kernel<<<...>>>( device_span{v.data(), v.size()} );
```

It also needs to guard against this:
```
thrust::device_vector<int> const const_v;
device_span<int> s(const_v); // this shouldn't work
device_span<const int> s(const_v); // this is okay
```

Reply:
Those cases are accounted for. :+1:

Reply:
I landed on using type traits to pick which containers each host/device span supports. This should prevent most accidental host/device mixups, but is very much a subset of `std::span`'s functionality.

const-ness is preserved.` Container<int> const` => `span<int const>`